src/core/thread/indirect_sender.cpp

```diff
@@ -520,6 +541,103 @@ void IndirectSender::ClearMessagesForRemovedChildren(void)
     }
 }
 
+void IndirectSender::ConvertChildMessagesToIndirect(Child &aChild)
```

Comment:
I wonder if we should be worried about the non-sleepy to sleepy case.
1. Queuing delays for direct transmissions should be relatively small, especially when compared to indirect transmissions.
1. A device may receive a given message multiple times (first as direct, then as indirect). The cycle may continue if the device continues to transition back and forth.
1. A message could get "pinned" in the transmit queue if multiple children are transitioning between sleepy and non-sleepy, and those transitions are out-of-phase.
1. This is a relatively large chunk of code for what I think is a relatively minor problem.

Thoughts?

Reply:
I agree...I was considering this actually (as I realized the logic to convert direct to indirect was getting more and more complicated).

On non-sleepy to sleepy, we can let any direct msg already in queue for the child to be sent as direct (which would fail) and basically dropped. One concern I had about this was that the tx failure would lead to `IncrementLinkFailures()` and may cause us to kick the child out and then not queue any new msg for it. Checking the code in `UpdateNeighborOnSentFrame()` noticed that we do not check link failures on a child. So this should not happen.

Will change the commits to only handle the sleepy to non-sleepy case.